IPNI_FEDERATION.md

```diff
@@ -0,0 +1,408 @@
+its own. Discrepancies are flagged based on a set of predefined conflict resolution rules. For instance, if two
+instances have added different data for the same key, the system might prioritize data from the instance with a more
+recent vector clock timestamp.
+
+Reconciliation is not just about resolving conflicts. It's also about merging updates. If an instance discovers newer
+data from a snapshot, it updates its index accordingly. Through continuous reconciliation processes, the protocol
+ensures that all IPNI instances trend towards eventual consistency.
+
+### Conflict Resolution
+
+In the vast landscape of the Inter-Planetary Network Indexer (IPNI), advertisement chains serve as the core mechanism
+for content propagation. As multiple IPNI instances exist across the globe, the potential for discrepancies in the state
+of these chains is undeniable. It becomes essential, therefore, to devise strategies that can reconcile these
+discrepancies effectively.
+
+#### Case 1: A provider is unknown by another IPNI instance
+
+**Scenario:**  
+Instance A is aware of Provider P1 and its advertisement chain, but Instance B is unaware of Provider P1 entirely.
+
+**Resolution Strategy:**  
+When instances exchange information, and Instance B identifies a missing provider, it should:
+
+1. Request the complete advertisement chain of Provider P1 from Instance A.
+2. Incorporate Provider P1's advertisement chain into its local index.
+3. Broadcast this new knowledge to other connected IPNI instances to ensure widespread awareness and data consistency.
+
+#### Case 2: A provider is known by both with a consistent head advertisement
+
+**Scenario:**  
+Both Instance A and Instance B are aware of Provider P1 and have the same head (latest) advertisement.
+
+**Resolution Strategy:**  
+This represents the ideal state of eventual consistency. In this scenario:
+
+1. Both instances recognize that they have the same version of the advertisement chain for Provider P1.
+2. No reconciliation action is required.
+3. The instances can continue to propagate other advertisement chains or updates, knowing that, for Provider P1, they
+   are in sync.
+
+#### Case 3: A provider is known by both but with different head advertisements
+
+**Scenario:**  
+Instance A and Instance B both know Provider P1, but they have different latest advertisements (heads) for P1's chain.
+
+**Resolution Strategy:**  
+This represents a divergence in the advertisement chain, and careful reconciliation is required:
+
+1. Instance A and B determine which advertisement is the later one in the chain of advertisements provided by P1.
+2. They converge on the advertisement has been published the latest, i.e. has the shorter distance to P1's current
+   advertisement head.
+
+While traversing advertisement chains, instances should prefer pulling ad chain from eachother's mirror instead of going
+directly to the provider where possible.
```

Comment:
This seems unfair to the faster indexer, because all the other indexers will pull from the mirror of the fastest. Maybe we need a strategy that strongly encourages other indexers to add content from others' mirrors to their own.

For example, an indexer cannot pull advertisement data directly from another mirror, but instead must sync some portion of the mirror and then get the advertisements from the local mirror after syncing. This forces the receiving indexer to have a copy of the data in its own mirror.

The original source indexer may then redirect others, for a short time, to the destination indexer for the synced content. Or, maybe there is some check that if one indexer copies advertisement data from another indexer's mirror, the indexer that received the data subsequently serves it from its mirror.

Reply:
In think in practice, the fastest indexer would/should have rate limiting in place, and other indexers should scatter traffic across multiple sources simply to pull ads from different providers in parallel.

The key point I am hoping to get across with this sentence is to say: try going to mirrors first before going to providers. That doesn't mean the same mirror all the time. 

The "fastest indexer" would also depend on where the providers and indexers physically are. I expect that the indexer network would naturally grow around the organisational structure of its providers. For example, if there is a heavy presence of providers in Australia , it is likely that there would be more than one indexer in Australia to cope with the demand. At this point I find myself heavily speculating which is probably a hint that we might be over thinking it.

Another factor here is the loosely defined concept of Mirror itself: there is no specification yet that documents the API for a mirror. We are in the process of figuring out its economics, rate limits etc. 

So i think we would probably be OK revisiting this once there is more data.

Thoughts?

Reply:
Is there anything that encourages one indexer to help another, when those indexers are not operated by the same organization? I guess just that it allows the other indexer to supply the ad data and reduces load on the first.

In other blockchain indexing ecosystems, the indexers are interested in handling as many queries as they can for the data they index - so they do not want to help other indexers. In the IPNI ecosystem, it appears that indexers are interested in doing as little of the required work as possible - so want other indexers to handle as many queries as possible.

Reply:
For now, I am OK with allowing indexers to share their mirror as much or as little as possible, with whatever limits they see fit, and specifying this a some later time.

Reply:
Captured this https://github.com/ipni/specs/issues/29